lib/spack/spack/binary_distribution.py

```diff
@@ -0,0 +1,209 @@
+
+from architecture import get_full_system_from_platform
+from spack.util.executable import which
+import spack.cmd
+import spack
+from spack.stage import Stage
+import spack.fetch_strategy as fs
+import spack.relocate
+
+
+def prepare():
+    """
+    Install patchelf as pre-requisite to the
+    required relocation of binary packages
+    """
+    if platform.system() == 'Darwin':
+        return
+    dir = os.getcwd()
+    patchelf_spec = spack.cmd.parse_specs("patchelf", concretize=True)[0]
+    if not spack.install_layout.check_installed(patchelf_spec):
+        patchelf = spack.repo.get(patchelf_spec)
+        patchelf.do_install()
+    os.chdir(dir)
+
+
+def buildinfo_file_name(spec):
+    """
+    Filename of the binary package meta-data file
+    """
+    return os.path.join(spec.prefix, ".spack", "binary_distribution")
+
+
+def read_buildinfo_file(package):
+    """
+    Read buildinfo file
+    """
+    filename = buildinfo_file_name(package)
+    with open(filename, 'r') as inputfile:
+        content = inputfile.read()
+        buildinfo = yaml.load(content)
+    return buildinfo
+
+
+def write_buildinfo_file(spec):
+    """
+    Create a cache file containing information
+    required for the relocation
+    """
+    text_to_relocate   = []
+    binary_to_relocate = []
+    blacklist = (".spack", "man")
+    for root, dirs, files in os.walk(spec.prefix, topdown=True):
+        dirs[:] = [d for d in dirs if d not in blacklist]
+        for filename in files:
+            path_name = os.path.join(root, filename)
+            filetype = spack.relocate.get_filetype(path_name)
+            if spack.relocate.needs_binary_relocation(filetype):
+                rel_path_name = os.path.relpath(path_name, spec.prefix)
+                binary_to_relocate.append(rel_path_name)
+            elif spack.relocate.needs_text_relocation(filetype):
+                rel_path_name = os.path.relpath(path_name, spec.prefix)
+                text_to_relocate.append(rel_path_name)
+
+    # Create buildinfo data and write it to disk
+    buildinfo = {}
+    buildinfo['buildpath'] = spack.install_path
+    buildinfo['relocate_textfiles'] = text_to_relocate
+    buildinfo['relocate_binaries']  = binary_to_relocate
+    filename = buildinfo_file_name(spec)
+    with open(filename, 'w') as outfile:
+        outfile.write(yaml.dump(buildinfo, default_flow_style=True))
+
+
+def tarball_directory_name(spec):
+    """
+    Return name of the tarball directory according to the convention
+    <os>-<architecture>/<compiler>/<package>/
+    """
+    return "%s/%s/%s" % (get_full_system_from_platform(),
+                         str(spec.compiler).replace("@", "-"),
+                         spec.name)
+
+
+def tarball_name(spec):
+    """
+    Return the name of the tarfile according to the convention
+    <os>-<architecture>-<package>-<dag_hash>.tar.gz
+    """
+    return "%s-%s-%s-%s.tar.gz" % (get_full_system_from_platform(),
```

Comment:
This should use `spack.architecture.sys_type()`, which contains the "platform", OS, and target... Any reason the package isn't first?  What about:

`<package>-<platform>-<os>-<target>-<dag_hash>.tar.gz`